Certs_check, derive commands (without verification), + some other changes

brh28 opened the pull request:
- Cert_check verifies the card against the Factory public key (Tapsigner only)
- `from_status` constructor for tapsigner
- Reorg of commands.rs to group request and responses by command type

notmandatory commented on src/pcsc.rs:
```diff
@@ -40,36 +40,22 @@ impl Transport for PcscTransport {
         let rapdu = transport.transmit_apdu(applet_select_apdu)?;
         let status_response = StatusResponse::from_cbor(rapdu.to_vec())?;
 
-        // get common fields
```
Is there a reason to duplicate these lines below instead of leaving here?

brh28 replied:
I think it would be cleaner to unpack the status response within the card types constructor.  I've been primarily working with the tapsigner, so that was the only constructor I added. This doesn't need to be in this PR, or if it is I should update for all three card types

notmandatory replied:
Ah ok I see now this is related to your `TapSigner::from_status` change. No need to change it.

notmandatory commented on src/lib.rs:
```diff
@@ -132,6 +156,78 @@ impl<T: Transport + Sized> TapSigner<T> {
+            card_nonce, 
+            nonce.clone(), 
+            &self.secp
+        ).is_ok());
```
Instead of an `assert!` how about adding new Error type `CertCheck(String)` and use it here like:
```suggestion
        ).map_err(|e| Error::CertCheck(e.to_string()))?;
```

Even though this is a fatal issue, we if just crash the app we don't have a chance to handle the error and tell the user what's going on.

notmandatory commented on src/lib.rs:
```diff
@@ -132,6 +156,78 @@ impl<T: Transport + Sized> TapSigner<T> {
+            pubkey = self.secp.recover_ecdsa(&md, &rec_sig).unwrap();
+        }
+
+        // todo - as a constant?
```
Here's an idea for how to use const values for this. But for this PR I don't mind leaving it as is so we can think about it some more. 

```rust
const PROD_FACTORY_ROOT_KEY:&str = "03028a0e89e70d0ec0d932053a89ab1da7d9182bdc6d2f03e706ee99517d05d9e1";
const TEST_FACTORY_ROOT_KEY:&str = "027722ef208e681bac05f1b4b3cc478d6bf353ac9a09ff0c843430138f65c27bab";

pub enum FactoryRootKey {
    Prod(PublicKey),
    Test(PublicKey),
}

impl TryFrom<PublicKey> for FactoryRootKey {
    type Error = Error;

    fn try_from(pubkey: PublicKey) -> Result<Self, Self::Error> {
        if pubkey.serialize().to_hex().eq(PROD_FACTORY_ROOT_KEY) {
            Ok(FactoryRootKey::Prod(pubkey))
        } else if pubkey.serialize().to_hex().eq(TEST_FACTORY_ROOT_KEY) {
            Ok(FactoryRootKey::Test(pubkey))
        } else {
            Err(Error::CertCheck(
                "Root cert is not from Coinkite. Card is counterfeit.".to_string(),
            ))
        }
    }
}

impl Debug for FactoryRootKey {
    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
        match &self {
            FactoryRootKey::Prod(pk) => {
                write!(f, "FactoryRootKey::Prod({:?})", pk)
            }
            FactoryRootKey::Test(pk) => {
                write!(f, "FactoryRootKey::Test({:?})", pk)
            }
        }
    }
}
```

notmandatory commented on src/lib.rs:
```diff
@@ -202,6 +309,86 @@ impl<T: Transport + Sized> SatsCard<T> {
+            Ok(value.to_string())
+        } else {
+            Err(Error::CiborValue("Root cert is not from Coinkite. Card is counterfeit.".to_string()))
+        }
```
If we go with a `FactoryRootKey` the end of this function becomes:
```suggestion
       pubkey.try_into()
```
And returns `Result<FactoryRootKey,Error>`.  But as mentioned above we don't need to decide for this PR, so unless you love my `Enum` idea we can leave this as is for now.

notmandatory reviewed (changes requested):
Great job figuring out how to verify the root certs! I also like the command/response reorg. Three things that I'd like you to do before merging are:
1. replace the `assert!` with an `Err()` in `certs_check`
2. remove replaced commented out code
3. be sure to do a do a `cargo fmt`

Thanks!

notmandatory commented:
I also tested with my TapSigner and SATsCard and cert verification works for me.

brh28 commented:
Added the FactoryRootKey struct as you suggested, and added a "name" function to return a displayable string.

Combined the certs_check into the SharedCommands. The only logic difference between cards is the message_bytes, which is defined in the "signed_message" function. I don't have an older versioned SatsCard (0.9.0), so haven't been able to test that condition.

Created an IncorrectSignature type for the Error enum. Would you prefer an error type for each command? (i.e rename to CertsCheck)

notmandatory reviewed (approved):
ACK f8bf6217f3dac629d0e42cb9312897efdc50f238

Looks good, but we're both going to have to be better about making cleaner commit histories on future PRs. But that should be easier with recent code reorgs behind us.

brh28 commented:
> ACK f8bf6217f3dac629d0e42cb9312897efdc50f238
> 
> 
> 
> Looks good, but we're both going to have to be better about making cleaner commit histories on future PRs. But that should be easier with recent code reorgs behind us.


Sorry about that, I usually use the "Squash & Merge" option on GitHub, so I have a bad habit about cleaning my commit history.
